### ISSUE TITLE
Cache cannot find the same object that we already have from a different query

### ISSUE THREAD

**Issue by kroussevrb** [apollographql/apollo-ios#2459]
## Bug report

Using the latest stable version of Apollo (0.53) with in-memory cache

`apollo.cacheKeyForObject = { $0["id"] }`

We perform this query when the user loads the app and show a list of products
```
query Products($accountId: ID!) {
  account(id: $accountId) {
    id
    onSale: products(first: 20, section: ON_SALE) {
      ...ProductPage
    }
    regular: products(first: 20, section: REGULAR) {
      totalCount
      ...ProductPage
    }
  }
}

fragment ProductPage on ProductConnection {
  pageInfo { hasNextPage endCursor }
  nodes { ...ProductFragment }
}
```

Now the user taps on one of the products
We start watching the product using `client.watch(GetProductQuery(productId: "asdasd", .returnCacheDataElseFetch)`

```
query GetProduct($productId: ID!) {
  product(id: $productId) {
    ...ProductFragment
  }
}
```

or 

```
query GetProduct($productId: ID!) {
  node(id: $productId) {
    ... on Product {
      ...ProductFragment
    }
  }
}
```

which is basically the same. `Product` implements `Node`

_Expected_: entry found in cache, no network request
_Actual_: product cache miss, network request performed for `GetProductQuery`. Consecutive taps on the product now find the product in the cache

**Details**
In `ApolloStore` when the following line of code is executed:
`return object[info.cacheKeyForField]`
when we print `object` and `info` we get

```
▿ 5 elements
  ▿ 0 : 2 elements
    - key : "myWishlist"
    ▿ value : -> #QUERY_ROOT.myWishlist
      - key : "QUERY_ROOT.myWishlist"
  ▿ 1 : 2 elements
    - key : "currencies"
    ▿ value : 172 elements
      ▿ 0 : -> #QUERY_ROOT.currencies.0
        - key : "QUERY_ROOT.currencies.0"
      ▿ 1 : -> #QUERY_ROOT.currencies.1
        …
        - key : "QUERY_ROOT.currencies.171"
  ▿ 2 : 2 elements
    - key : "viewer"
    ▿ value : -> #qweqwe
      - key : "qweqwe"
  ▿ 3 : 2 elements
    - key : "account(id:asdasd)" <---- this, if expanded, contains "products"
    ▿ value : -> #asdasd
      - key : "asdasd"
  ▿ 4 : 2 elements
    - key : "myAccount"
    ▿ value : -> #asdasd
      - key : "asdsd"
```

however, `info.cacheKeyForField` is "product(id:qweqwe)"
and when the `object` is subscripted for it, it returns `nil` without checking for the nested paths and without looking up a flat hierarchy of entities, where this object will be present by its id (evident by using `readFragment` function where we pass ProductFragment.self and the id and we get it from there)

Once the user taps on the product and `GetProductQuery` is performed, and we once again go through
`return object[info.cacheKeyForField]`

if we print `object` again this time, it returns the same list of entries **plus a new one**

```
▿ 6 elements
  ▿ 0 : 2 elements
    - key : "myWishlist"
    ▿ value : -> #QUERY_ROOT.myWishlist
      - key : "QUERY_ROOT.myWishlist"
  ▿ 1 : 2 elements
    - key : "currencies"
    ▿ value : 172 elements
      ▿ 0 : -> #QUERY_ROOT.currencies.0
        - key : "QUERY_ROOT.currencies.0"
      ▿ 1 : -> #QUERY_ROOT.currencies.1
        …
        - key : "QUERY_ROOT.currencies.171"
  ▿ 2 : 2 elements
    - key : "viewer"
    ▿ value : -> #qweqwe
      - key : "qweqwe"
  ▿ 3 : 2 elements
    - key : "account(id:asdasd)" <---- this, if expanded, contains "products"
    ▿ value : -> #asdasd
      - key : "asdasd"
  ▿ 4 : 2 elements
    - key : "myAccount"
    ▿ value : -> #asdasd
      - key : "asdsd"
   ▿ 5 : 2 elements
    - key : "product(id:qweqwe)" <-- from here our Product is now accessible although we've had it all along
    ▿ value : -> #qweqwe
      - key : "qweqwe"
```

and now subcripting `object` with `info.cacheKeyForField` which is "product(id:qweqwe:" finds the result

Basically we had the product from the get go but it was nested under `account` query. The lookup neither expanded to check what's in account nor looked in a flattened table of entries in the graph to find it there.

The same thing works on Android.

## Versions
- `apollo-ios` SDK version: 0.53
- Xcode version: 13.4.1
- Swift version: 5.x
- Package manager: SPM

@calvincestari

**Comment by AnthonyMDev** [comment]
This does not work as expected because the client can't know that `product(id: "qweqwe")` is going to return the `Product` with the `id` of `qweqwe`. That sounds unintuitive, but really, this is a server-side implementation detail. All we know is that you have a field that takes an argument, we don't know what the server does with that argument or what it plans on returning.

Apollo Kotlin has a concept of [`fieldPolicy` directives](https://www.apollographql.com/docs/kotlin/caching/declarative-ids/#fieldpolicy). This allows you to tell the client, "For this field, the server will return an object of the fields return type with a cache key equal to the argument `id` value."  If this is working on Android, then you must have a field policy set up there.

Unfortunately, field policies aren't a thing in Apollo iOS yet. But they are awesome and we fully intend to implement something like this on iOS in the future. We've been working hard on getting the 1.0 release ready, which provides a much more flexible API for us to extend into features such as this in the future.

In the mean time, the work around would be to do a cache read for your `ProductFragment` as you explained above, and if you don't get an object returned from that, then send the query to the network.

I highly recommend you migrate your application to the 1.0 version as soon as you can, as all new features are going to be added to that version going forward. The beta is getting pretty close to stable now, but documentation and migration guidance is lacking. We aim to have an official release by the end of September!
